text/0000-default-field-values.md

```diff
@@ -0,0 +1,1767 @@
+
+4. For each field `f` in `τ` *not* mentioned in `fields`, a default value must
+   exist for `f` in the `struct` or `enum` variant referenced by `path`.
+
+---------------------
+
+**Lemma 1:** If `τ` is of a type `σ` where `σ: Sync`, and if all `field: expr`
+has an `expr` component which is a valid constant expression, then `τ` is
+also a valid constant expression.
+
+**Proof:** It trivially holds that if all `expr`s in `τ` are
+constant expressions and all default values for fields omitted
+are also constant expressions, then `τ` is a constant expression.
+
+**Lemma 2:** If an expression `x` of form `Path { .. }` is of a type `σ` where
+`σ: Sync`, then `x` is a constant expression.
+
+**Proof:** `x` is a special case of `τ` with all fields according to the default
+values. Since there are no explicitly provided fields, by Lemma 1 it holds.
+
+### Dynamic semantics
+
+Given a well-formed expression `τ` according to the aforementioned form,
+the operational semantics is the same as the semantics of an expression
+`path { new_fields }` with `new_fields = fields ∪ default_fields` where:
+
++ `∪` computes the left-biased union where equal elements are determined
+  solely on equality of field names.
+
++ `fields` are the set of explicitly given `field: expr` pairs.
+
++ `default_fields` are the fields in `τ`'s data-type or variant which
+  have provided default values.
+
+## `#[derive(Default)]`
+
+When generating an implementation of `Default` for a `struct` named `$s` on
+which `#[derive(Default)]` has been attached, the compiler will omit all fields
+which have default values provided in the `struct`. The the associated function
+`default` shall then be defined as (where `$f_i` denotes a vector of the fields
+of `$s`):
+
+```rust
+fn default() -> Self {
+    $s { $f_i: Default::default(), .. }
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The usual drawback of increasing the complexity of the language applies.
+However, the degree to which complexity is increased is not substantial.
+
+In particular, the syntax `Foo { .. }` mirrors the identical and already
+existing pattern syntax. This makes the addition of `Foo { .. }` at worst
+low-cost and potentially cost-free.
+
+The main complexity comes instead from introducing `field: Type = expr`.
+However, as seen in the [prior-art], there are several widely-used languages
+that have a notion of field / property / instance-variable defaults.
+Therefore, the addition is intuitive and thus the cost is seen as limited. As
+an implementation detail, `rustc` *already* parses `field: Type = expr`
+purely to provide an appropriate diagnostic error:
+
+```
+error: default values on `struct` fields aren't supported
+ --> src/lib.rs:2:28
+  |
+2 |     pub alpha: &'static str = "",
+  |                            ^^^^^ help: remove this unsupported default value
+```
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
```

Comment:
I would like to see a section on differences from FRU (especially the interaction with privacy and `#[non_exhaustive]`) including why it is the way it is and the reason for diverging from it in this RFC. And possibly a future possibility about how they can be made more consistent in the future.

I'm generally a big fan of this RFC, but undoing past mistakes by adding more features without fixing the features we have leads to an uneven and complex language surface. So let's try to make sure we can do both.

Reply:
I think that it is reasonable to change this RFC to explicitly follow RFC-0736, and disallow the construction of types with private fields, if that will make this RFC less controversial. Particularly when considering the interaction with `#[derive(Default)]`, the feature is useful on its own without expanding the field privacy rules.

Reply:
That sketch uses a completely new syntax, so we wouldn't be fixing FRU so much as deprecating it and replacing it with something else.

It seems better to design this one to work the way we want from the beginning? Especially since people will have to opt in and might rely on restrictions we implement today (even if we document that those restrictions will be lifted in the future.. people don't usually _read the docs_ on language features so much as _try them out in the playground_).